cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -158,15 +184,29 @@ public void deleteApplications(String namespaceId) {
   public void updateAppSpec(String namespaceId, String appId, String versionId, ApplicationSpecification spec) {
     LOG.trace("App spec to be updated: id: {}: spec: {}", appId, GSON.toJson(spec));
     MDSKey key = new MDSKey.Builder().add(TYPE_APP_META, namespaceId, appId, versionId).build();
+    MDSKey versionLessKey = null;
     ApplicationMeta existing = getFirst(key, ApplicationMeta.class);
+    ApplicationMeta updated;
+
+    // Check again without the version to account for old data format if might not have been upgraded yet
+    if (existing == null && (versionId.equals(ApplicationId.DEFAULT_VERSION))) {
```

Comment:
In that case, the write will replace the existing appSpec which is already versioned. Why do we need to delete the versionLessKey in that case? The first time an app is created with default version (and the same app exists without version), writeApplication method will delete the old versionLessone and add a new one with version.